### ISSUE TITLE
Criticality score should include dependency information

### ISSUE THREAD

**Issue by Zac-HD** [ossf/criticality_score#8]
👋 Hi there!  I've recently gotten involved with OSSF based on my work on testing and fuzzing Python - I help maintain Pytest and am a core dev of Hypothesis (both in the current top-50), among other packages.

While I really like the idea of a criticality score, the current equation is almost entirely based on popularity, in the sense of "projects that people know about and interact with a lot"... but some of the most serious risks IMO come from projects that *everyone depends on but few people know about*.  The 'mentions' part would at least pick up OpenSSL, but conceptually I think that `A depends on B` implies `criticality(A) < criticality(B)` - for example, Pytest's dependencies are at least as widely used as Pytest itself, and therefore the impact of a vulnerability or malicious patch would be even more serious.

Specifically for Python, there's a lot of work involved in getting a reliable dependency graph - see https://github.com/pypa/packaging-problems/issues/54 - and accelerating that is probably a [fundable packaging improvement](https://github.com/psf/fundable-packaging-improvements).

TLDR: it's supply-chain attacks all the way down 😬

**Comment by inferno-chromium** [comment]
It does it using "commit mentions" [https://github.com/ossf/criticality_score/blob/06f6598c9eefa094c2c530bcb3919ae34fa6b844/criticality_score/run.py#L142] with a high weight, which seems to work across all languages, including c/c++. i was looking at github graphql api too, but current dependency information only works for those with package managers [https://github.com/tensorflow/tensorflow/network/dependents]. PRs welcome on any other ideas you have.

**Comment by Zac-HD** [comment]
Yep, the 'mentions' part of the equation is both elegant and important, but I don't think it captures everything - for example, `pytest-xdist` depends on [`execnet`](https://pypi.org/project/execnet/) to handle the distributed (i.e. "execute arbitrary code received over the network"!) parts, but [it has less than half](https://github.com/search?q=execnet) the [github search results](https://github.com/search?q=pytest-xdist) and is no longer updated.

*If* we had a reliable dependency graph, we could calculate a dependency-aware criticality score as e.g. `max(score, score/2 + max_score_of_any_dependent_package)` - this is still a little janky, but I think looking at packages where the dependency-graph-aware score is much higher than the existing score would reveal some under-maintained packages.

Maybe we're also thinking about slightly different problems - the existing score is a great way to look at active and high-engagement projects, whereas I'm concerned about projects that are inactive or ignored but still widely used.

**Comment by inferno-chromium** [comment]
> Yep, the 'mentions' part of the equation is both elegant and important, but I don't think it captures everything - for example, `pytest-xdist` depends on [`execnet`](https://pypi.org/project/execnet/) to handle the distributed (i.e. "execute arbitrary code received over the network"!) parts, but [it has less than half](https://github.com/search?q=execnet) the [github search results](https://github.com/search?q=pytest-xdist) and is no longer updated.
> 
> _If_ we had a reliable dependency graph, we could calculate a dependency-aware criticality score as e.g. `max(score, score/2 + max_score_of_any_dependent_package)` - this is still a little janky, but I think looking at packages where the dependency-graph-aware score is much higher than the existing score would reveal some under-maintained packages.
> 
> Maybe we're also thinking about slightly different problems - the existing score is a great way to look at active and high-engagement projects, whereas I'm concerned about projects that are inactive or ignored but still widely used.

Yes i agree with you here and want to find a way to take these into account. We still provide the --params field so that orgs can use internal usage info on libs and re-adjust score. But i do want to find a way to account for these in general.
